lisa/util/parallel.py

```diff
@@ -1,11 +1,65 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT license.
-import concurrent.futures
+
 import time
-from typing import Any, Callable, List, Optional
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Any, Callable, Generic, List, Optional, TypeVar
 
+from . import LisaException
 from .logger import Logger
 
+T_RESULT = TypeVar("T_RESULT")
+
+
+class TaskManager(Generic[T_RESULT]):
+    def __init__(self, max_workers: int, callback: Callable[[T_RESULT], None]) -> None:
+        self._pool = ThreadPoolExecutor(max_workers=max_workers)
+        self._max_workers = max_workers
+        self._futures: List[Future[T_RESULT]] = list()
+        self._callback = callback
+        self._cancelled = False
+
+    def __enter__(self) -> Any:
+        return self._pool.__enter__()
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> Optional[bool]:
+        return self._pool.__exit__(exc_type, exc_val, exc_tb)
+
+    def submit_task(self, task: Callable[[], T_RESULT]) -> None:
+        future: Future[T_RESULT] = self._pool.submit(task)
+        self._futures.append(future)
+
+    def cancel(self) -> None:
+        self._cancelled = True
+
+    def check_cancelled(self) -> None:
+        if self._cancelled:
+            raise LisaException("Tasks are cancelled")
+
+    def has_idle_worker(self) -> bool:
+        return len(self._futures) < self._max_workers
+
+    def poll(self) -> bool:
+        """
+        Return:
+            True, if there is idle worker.
+        """
+
+        # if there is any change, skip sleep to get faster
+        changed = False
+        for future in self._futures:
+            # join exceptions of subthreads to main thread
+            if future.done():
+                changed = True
+                # removed finished threads
+                self._futures.remove(future)
+                # exception will throw at this point
+                self._callback(future.result())
+                break
+        if not changed:
+            time.sleep(0.01)
```

Comment:
Can't we use condition objects/variables to have a sleep-less implementation?

Reply:
sleep-less run on CPU usually, it causes one CPU looks fully used. If it's short time like spin locker, it should be ok. But here is long time running, I hope it release CPU cycles more. So use a short sleep here, it's about 10ms.

Reply:
Not at all. To save me from typing here, check https://stackoverflow.com/questions/65147546/understanding-condition-variables-isnt-there-always-spin-waiting-somewhere. You put the wait at OS-level with that path, not a spin on userland. With your code, you hijack that CPU much harder, like you said, just to check on the siblings.

Reply:
The current solution has no locker involved, and the logic is pretty simple. The time of sleep doesn't need to be accurate, it makes sure wait at least 10 ms. After 10 ms, it's the same as other thread is scheduled. The alternative way is to call `future.result` with a timeout parameter, but it needs to wait one by one. It means there may be up to 30 (in smoke test) futures to be looped, and lock/unlock frequently. Either it's run frequently to get main thread respond quickly with more resource, or run it timeout longer to save resource, but much slower response in main thread.

Let me know, if I misunderstand it.

Reply:
It's simple, but CPU-intensive. Why not use ThreadPoolExecutor's https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.as_completed thing, instead of looping ourselves--that's one alternative, no?

Reply:
CPU-intensive depends on cycle of sleeps, the CPU cost is ignorable with 10ms. Anyway, changed to use `wait` for futures. It uses a chain of waiter, not sure it's efficiency on 30 threads.

Reply:
Awesome, thanks. That seems to be equivalent to as_completed (with that args), and it seems to do the right thing:
    waiter.event.wait(timeout)
    for f in fs:
        with f._condition:
            f._waiters.remove(waiter)
It should perform just fine, while being way gentler to the CPU with less activitiy.